src/libraries/System.Private.DataContractSerialization/src/System/Runtime/Serialization/EnumDataContract.cs

```diff
@@ -80,42 +83,59 @@ private sealed class EnumDataContractCriticalHelper : DataContract.DataContractC
-                Add(typeof(int), "int");
-                Add(typeof(uint), "unsignedInt");
-                Add(typeof(long), "long");
-                Add(typeof(ulong), "unsignedLong");
+                Add(typeof(sbyte), DictionaryGlobals.SignedByteLocalName.Value);        // "byte"
+                Add(typeof(byte), DictionaryGlobals.UnsignedByteLocalName.Value);       // "unsignedByte"
+                Add(typeof(short), DictionaryGlobals.ShortLocalName.Value);             // "short"
+                Add(typeof(ushort), DictionaryGlobals.UnsignedShortLocalName.Value);    // "unsignedShort"
+                Add(typeof(int), DictionaryGlobals.IntLocalName.Value);                 // "int"
+                Add(typeof(uint), DictionaryGlobals.UnsignedIntLocalName.Value);        // "unsignedInt"
+                Add(typeof(long), DictionaryGlobals.LongLocalName.Value);               // "long"
+                Add(typeof(ulong), DictionaryGlobals.UnsignedLongLocalName.Value);      // "unsignedLong"
             }
 
             internal static void Add(Type type, string localName)
             {
-                XmlQualifiedName stableName = CreateQualifiedName(localName, Globals.SchemaNamespace);
-                s_typeToName.Add(type, stableName);
-                s_nameToType.Add(stableName, type);
+                XmlQualifiedName xmlName = CreateQualifiedName(localName, Globals.SchemaNamespace);
+                s_typeToName.Add(type, xmlName);
+                s_nameToType.Add(xmlName, type);
+            }
+
+            internal static XmlQualifiedName GetBaseContractName(Type type)
+            {
+                s_typeToName.TryGetValue(type, out XmlQualifiedName? retVal);
+
+                Debug.Assert(retVal != null);   // Enums can only have certain base types. We shouldn't come up empty here.
+                return retVal;
+            }
+
+            internal static Type? GetBaseType(XmlQualifiedName baseContractName)
+            {
+                s_nameToType.TryGetValue(baseContractName, out Type? retVal);
+                return retVal;
             }
 
             [RequiresUnreferencedCode(DataContract.SerializerTrimmerWarning)]
             internal EnumDataContractCriticalHelper(
                 [DynamicallyAccessedMembers(ClassDataContract.DataContractPreserveMemberTypes)]
                 Type type) : base(type)
             {
-                this.StableName = DataContract.GetStableName(type, out _hasDataContract);
+                XmlName = DataContract.GetXmlName(type, out _hasDataContract);
                 Type baseType = Enum.GetUnderlyingType(type);
+                XmlQualifiedName baseTypeName = GetBaseContractName(baseType);
+                _baseContract = DataContract.GetBuiltInDataContract(baseTypeName.Name, baseTypeName.Namespace)!;
+                // NOTE TODO smolloy - Setting XmlName might be redundant. But I don't want to miss an edge case.
+                _baseContract.XmlName = baseTypeName;
```

Comment:
A simple setter on a non-hot code path is totally fine compromise to make sure nothing was missed.